cpp/include/partition_manager.hpp

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/comms/comms.hpp>
+#include <raft/handle.hpp>
+
+#include <cassert>
+#include <cmath>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <vector>
+
+namespace cugraph {
+namespace partition_2d {
+template <typename from_t>
+std::string to_string(from_t const& value)
+{
+  std::stringstream ss;
+  ss << value;
+  return ss.str();
+}
+
+// class responsible for creating 2D partition of workers:
+// responsible with finding appropriate P_ROW x P_COL
+// 2D partition and initializing the raft::handle_t communicator
+//
+// (this might be removed; or, it might exist already)
+//
+template <typename size_type = int>
+class partition_manager_t {
+ public:
+  partition_manager_t(raft::handle_t& handle, size_type p_row_size, size_type p_col_size)
+    : handle_(handle), p_row_size_(p_row_size), p_col_size_(p_col_size)
+  {
+    init_communicator();
+  }
+
+  partition_manager_t(raft::handle_t const& handle, size_type p_size) : handle_(handle)
+  {
+    partition2d(p_size);
+    init_communicator();
+  }
+
+  virtual ~partition_manager_t(void) {}
+
+ protected:
+  virtual void partition2d(size_type p_size)
+  {
+    auto sqr = static_cast<size_type>(std::sqrt(p_size));
+
+    // find divisor of p_size
+    // nearest to sqr;
+    //
+    p_row_size_ = nearest_divisor(sqr, p_size);
+    p_col_size_ = p_size / p_row_size_;
+
+    assert(p_row_size_ > 1 && p_col_size_ > 1);
+  }
+
+  virtual void init_communicator(void)
+  {
+    // TODO: init's handle's communicator (singleton?)
+  }
+
+ private:
+  raft::handle_t& handle_;
```

Comment:
Once handle_ is initialized, we can retrieve p_row_size_ from handle_.get_subcomms(row_key).get_size(); I think p_row_size_ & p_col_size_ below are redundant, and not sure we need this class.

Reply:
In light of recent changes, we might _not_ need this class anymore. I kept it around because a previews comment said there are some constraints between the calls on communicators in a handle. 

As usually, I think such constraints / pre-conditions should be enforced (RAII-ed?) in code _somewhere_. Well, such a _somewhere_ can be this class.

Reply:
Removed partition_manager_t class as requested. Resolving.